server/hap2/haplib.py

```diff
@@ -304,3 +308,166 @@ def error(self, error_code, response_id):
 
     def notify(self, procedure_name, params):
         self.request(procedure_name, params, request_id=None)
+
+
+class Utils:
+    # TODO: We need to specify custom validators
+    # TODO: Check the maximum length
+    # ToDo Currently, this method does not have notification filter.
+    # If we implement notification procedures, should insert notification filter.
+    @staticmethod
+    def define_transporter_arguments(parser):
+        parser.add_argument("--transporter", type=str,
+                            default="RabbitMQHapiConnector")
+        parser.add_argument("--transporter-module", type=str, default="haplib")
+
+        # TODO: Don't specifiy a sub class of transporter directly.
+        #       We'd like to implement the mechanism that automatically
+        #       collects transporter's sub classes, loads them,
+        #       and calls their define_arguments().
+        RabbitMQHapiConnector.define_arguments(parser)
+
+    @staticmethod
+    def load_transporter(args):
+        (file, pathname, descr) = imp.find_module(args.transporter_module)
+        mod = imp.load_module("", file, pathname, descr)
+        transporter_class = eval("mod.%s" % args.transporter)
+        return transporter_class
+
+    @staticmethod
+    def parse_received_message(message, allowed_procedures):
+        """
+        Parse a received message.
+        @return A ParsedMessage object. Each attribute will be set as below.
+        - error_code If the parse is succeeded, this attribute is set to None.
+                     Othwerwise the error code is set.
+        - message_dict A dictionary that corresponds to the received message.
+                       If the parse fails, this attribute may be None.
+        - message_id A message ID if available. Or None.
+        - error_message An error message.
+        """
+
+        pm = ParsedMessage()
+        pm.error_code, pm.message_dict = \
+          Utils._convert_string_to_dict(message)
+
+        # Failed to convert the message to a dictionary
+        if pm.error_code is not None:
+            return pm
+
+        pm.message_id = pm.message_dict.get("id")
+
+        # The case the message is a reply
+        need_id = True
+        should_reply = pm.message_dict.has_key("result")
+
+        if not should_reply:
+            method = pm.message_dict["method"]
+            pm.error_code = Utils.is_allowed_procedure(method,
+                                                       allowed_procedures)
+            if pm.error_code is not None:
+                pm.error_message = "Unsupported method: '%s'" % method
+                return pm
+            if PROCEDURES_DEFS[method].get("notification"):
+                need_id = False
+
+        # 'id' is not included in the message.
+        if need_id and pm.message_id is None:
```

Comment:
If the message is an error response, message_id can be "null".
The "null" in JSON world is represented by "None" in Python world, isn' it?
This code can't distinguish whether message_id doesn't exist or message_id is null.